[Proposal] Danger rule for xcodeproj (#112)
Babylonpartners/ios-playbook: Cookbook/Proposals/Danger-Rule-Xcodeproj.md

# Danger Rules for pbxproj

* Author: Olivier Halligon
* Review Manager: TBD

## Introduction

This proposal suggests to add some Danger rules checking for `xcodeproj` consistenty.

It is dependent on [#110 - Integrate Danger](https://github.com/Babylonpartners/ios-playbook/pull/110) being accepted first.

## Motivation

`pbxproj` project format is hard to parse, which means it's easy to miss some mistakes that could happen when an Xcode project has changed during a PR.

This means that some mistakes have ended up in `develop` in the past, like:

> * `(null)` references in `pbxproj` files

Author: ```suggestion
* `(null)` references and `"Recovered References"` groups in `pbxproj` files
```

* Files being added as part of a target membership while they shouldn't:
  * Snapshot files
  * xcconfig files
  * Info.plist files

> Some of those mistakes, especially `(null)` references, are generally the result of build conflicts; so their apparition during a PR could be an indicator that something else could have gone wrong during the merge of the pbxproj file. It's generally a good incentive to double-check that there was no other side effect from the pbxproj merge.

Reviewer: ```suggestion
Some of those mistakes, especially `(null)` references, are generally the result of build conflicts; so their appearances during a PR could be an indicator that something else could have gone wrong during the merge of the pbxproj file. It's generally a good incentive to double-check that there was no other side effect from the pbxproj merge.
```

Author: 👍 Actually, even:
```suggestion
Some of those mistakes, especially `(null)` references, are generally the result of merge conflicts; so their appearances during a PR could be an indicator that something else could have gone wrong during the merge of the pbxproj file. It's generally a good incentive to double-check that there was no other side effect from the pbxproj merge.
```

> Some of those mistakes, especially `(null)` references, are generally the result of merge conflicts; so their appearances during a PR could be an indicator that something else could have gone wrong during the merge of the pbxproj file. It's generally a good incentive to double-check that there was no other side effect from the pbxproj merge.

Author: ```suggestion
Some of those mistakes, especially `(null)` and "Recovered References" in `pbxproj`, are generally the result of merge conflicts; so their appearances during a PR could be an indicator that something else could have gone wrong during the merge of the pbxproj file. It's generally a good incentive to double-check that there was no other side effect from the pbxproj merge.
```

Other mistakes like unexpected files in target membership could lead to files ending up in the final bundle uploaded to the AppStore, and in addition to being useless in the final `ipa`, bloat the app size for no reason.

## Proposed solution

> * We can use a simple "find" in every `*.pbxproj` for the string `/* (null) */`, and make Danger generate an inline comment on the found line(s) if any

Reviewer 2: Will Danger automatically provide timing values for all its tasks or its something that we would need to implement ourselves? I think it's something we can be interested in monitoring to avoid that those checks are taking a very long time considering the size of our project (actually same can be applied to other tools like swiftlint @diegopetrucci @TheAdamBorek )

Author: Good question.
I don't think Danger provides timing info. What we could imagine doing is get the time at the beginning and end of Dangerfile and include a `message('Danger took 160ms to run its rules')` at the very end.

Not sure it's worth it though tbh, as like it's explained in the main #110 proposal, as opposed to SwiftLint — which needs to go thru every file in the project (changed or not) one by one and parse their content and even analyze the SIL representation for some rules — Danger only do some tests on the PR metadata… or grep on a very small subset of files (like one or two pbxproj which have changed in the PR), so it's way less time-consuming compared to e.g. SwiftLint.

> * We can use a simple "find" in every `*.pbxproj` for the string `/* (null) */`, and make Danger generate an inline comment on the found line(s) if any

Author: ```suggestion
* We can use a simple "find" in every `*.pbxproj` for the strings `/* (null) */` and `name = "Recovered References";`, and make Danger generate an inline comment on the found line(s) if any
```

This can be done pretty easily with Ruby, conceptually this will look like the logic below:

```ruby
pbxprojs = git.modified_files.select { |f| f.end_with?('pbxproj') }
pbxprojs.each do |filename|
  File.foreach(filename).with_index do |line, line_num|

>     warn("null reference found in pbxproj", file: filename, line: line_num) if line.include?('/* (null) */') }

Author: ```suggestion
    warn("null reference found in pbxproj", file: filename, line: line_num) if line.include?('/* (null) */') }
    warn("recovered references found in pbxproj", file: filename, line: line_num) if line.include?('name = "Recovered References";') }
```

pbxprojs.each do |project_file|
  proj = Xcodeproj::Project.open(project_file)
  proj.targets.each do |target|
    rsrc_files = target.build_phases
      .find { |p| p.is_a?(Xcodeproj::Project::Object::PBXResourcesBuildPhase) }
      .files_references.map(&:path)
    rsrc_files.each do |path|
      fail("#{path} should be removed from #{project_file}:#{target.name}" if path.end_with? '.xcconfig'
    end
  end
end
```

* We can use a similar logic to warn about `Info.plist` files added to targets' resource phases by mistake.
* We can also use a similar rule to prevent `__Snapshots__` directories be added in resources... and even to be added as reference to any `xcodeproj` at all.

All those rules listed above, if implemented in the `Dangerfile`, will help us:

* Keep the Project files sanitised and avoid commiting invalid pbxproj files, which are often the result of merge conflicts.
* Be aware of potential higher risks of a merge gone wrong on a pbxproj file, to avoid commiting a corrupted pbxproj (e.g. one that could still be readable by `xcodebuild` but in which a reference to an image file got removed from the pbxproj due to a bad merge, making the app crash at runtime)

> * Avoid including large chunk of unrelevant resources (like test snapshots) in the final bundle

Author: - Testing that a file named `*Tests.swift` is included in the test target: should not be that complicated to write, I've written rules similar to that logic in the past and the xcodeproj gem is quite efficient for this kind of things.

- Checking it's also in the test action of the schemes: I have never tested writing such a rule for Danger; but from what I remember from when I contributed to the XCScheme parser years ago, we should have everything we need in the xcodeproj gem to write such a test as well, so I think it's totally doable.

---

What might be more "difficult" is for us to write down and define what the rules for such a check should be: on what criterion do we decide if a test file is in the *right* test target? what if a test file is in a test target but not the right one, or in one test action of one scheme but not the right one, what check could we imagine to detect that? Or do we consider that testing that the test file is "at least in one test target no matter which one" is a good enough test?

> * Avoid including large chunk of unrelevant resources (like test snapshots) in the final bundle

Author: Gave it a quick try and came up with this logic: https://gist.github.com/AliSoftware/447c5ed5e04d2f25f14d1922757048b7

> * Avoid including large chunk of unrelevant resources (like test snapshots) in the final bundle

Reviewer 2: I wouldn't complicate it. We just need to make sure that there are no tests which are not linked to any test target.

> * Avoid including large chunk of unrelevant resources (like test snapshots) in the final bundle

Reviewer 2: but that's totally not a priority, so please don't invest more effort into it!

> * Avoid including large chunk of unrelevant resources (like test snapshots) in the final bundle

Author: Yeah, I agree, I wasn't planning on adding that to the proposal yet anyway (don't worry I just toyed with it for like 5mn while waiting for a compilation to end in the background :P )
